src/bootstrap/bin/rustc.rs

```diff
@@ -106,6 +106,15 @@ fn main() {
         {
             cmd.arg("-C").arg("panic=abort");
         }
+
+        // `-Ztls-model=initial-exec` must not be applied to proc-macros, see
+        // issue https://github.com/rust-lang/rust/issues/100530
+        if env::var("RUSTC_TLS_MODEL_INITIAL_EXEC").is_ok()
+            && arg("--crate-type") != Some("proc-macro")
+            && !matches!(crate_name, Some("proc_macro2" | "quote" | "syn" | "synstructure"))
```

Comment:
Is there a test or something else we can add that validates this list is enough? For example, I'm thinking that maybe our invocations of cargo metadata can help here?

I don't think this is something we should do in rustc.rs, probably too much overhead, but some kind of out of band check seems like a good idea to me.

Reply:
I guess the actual set is something like proc macros dep *and* contains TLS symbols, or so?

For example, rustc_macros depends on multiple other crates:

https://github.com/rust-lang/rust/blob/master/compiler/rustc_macros/Cargo.toml#L10

Reply:
> I guess the actual set is something like proc macros dep and contains TLS symbols, or so?

Yes, although a blanket include of all proc macro deps would be correct though. It may just cause a perf regression.

Reply:
I think we should try it; my guess is that TLS access for these crates isn't hot (even at runtime if they're pulled in there).

Or we can go a different direction and try to test this, but it sounds like doing that isn't trivial?

Reply:
I think checking all proc macro dylibs shipped with rustc-dev for the R_X86_64_TPOFF64 relocation kind (on x86_64, aarch64 has a different relocation name) would suffice. R_X86_64_DTPOFF64 is the only TLS relocation kind that is used for the global dynamic tls model (the most general tls model) I think.

Reply:
The difference before/after appears to be R_AARCH64_TLS_TPREL64

Is there a nice way to get a list of the proc macro dylibs in a `run-make-fulldeps` test? (or a better way to test for that)

Reply:
Not sure.